File: Binary Tree/OrderTraversal.cpp
#include <bits/stdc++.h>
using namespace std;

// Node structure: left pointer, value, right pointer
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};

class Tree {
private:
    Node* root;

public:
    Tree() {
        root = NULL;
    }

    // Build Tree from space-separated level order string
    void BuildTree(string str) {
        if (str.empty() || str[0] == 'N') return;

        vector<string> info;
        stringstream ss(str);
        string word;
        while (ss >> word) {
            info.push_back(word);
        }

        int val = stoi(info[0]);
        root = new Node(val);

        queue<Node*> Q;
        Q.push(root);

        int i = 1;
        while (!Q.empty() && i < info.size()) {
            Node* curr = Q.front();
            Q.pop();

            // Left child
            if (i < info.size() && info[i] != "N") {
                val = stoi(info[i]);
                curr->left = new Node(val);
                Q.push(curr->left);
            }
            i++;

            // Right child
            if (i < info.size() && info[i] != "N") {
                val = stoi(info[i]);
                curr->right = new Node(val);
                Q.push(curr->right);
            }
            i++;
        }
    }

    // Inorder Traversal (Left, Root, Right)
    void inorder(Node* node) {
        if (!node) return;
        inorder(node->left);
        cout << node->data << " ";
        inorder(node->right);
    }

    // Preorder Traversal (Root, Left, Right)
    void preorder(Node* node) {
        if (!node) return;
        cout << node->data << " ";
        preorder(node->left);
        preorder(node->right);
    }

    // Postorder Traversal (Left, Right, Root)
    void postorder(Node* node) {
        if (!node) return;
        postorder(node->left);
        postorder(node->right);
        cout << node->data << " ";
    }

    // Public wrappers
    void printInorder() {
        cout << "Inorder Traversal: ";
        inorder(root);
        cout << endl;
    }

    void printPreorder() {
        cout << "Preorder Traversal: ";
        preorder(root);
        cout << endl;
    }

    void printPostorder() {
        cout << "Postorder Traversal: ";
        postorder(root);
        cout << endl;
    }
};

int main() {
    Tree t1;
    string str;
    cout << "Enter tree (level order, space separated): ";
    getline(cin, str); // Example: 10 20 30 N 40 N N

    t1.BuildTree(str);

    t1.printInorder();
    t1.printPreorder();
    t1.printPostorder();

    return 0;
}

File: Binary Tree/Order_Traversal.py
from collections import deque

# Node structure
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None

# Tree class
class Tree:
    def __init__(self):
        self.root = None

    # Build Tree from space-separated level order string
    def build_tree(self, s):
        if not s or s[0] == 'N':
            return

        info = s.split()
        self.root = Node(int(info[0]))

        queue = deque([self.root])
        i = 1

        while queue and i < len(info):
            curr = queue.popleft()

            # Left child
            if i < len(info) and info[i] != 'N':
                curr.left = Node(int(info[i]))
                queue.append(curr.left)
            i += 1

            # Right child
            if i < len(info) and info[i] != 'N':
                curr.right = Node(int(info[i]))
                queue.append(curr.right)
            i += 1

    # Inorder Traversal (Left, Root, Right)
    def inorder(self, node):
        if not node:
            return
        self.inorder(node.left)
        print(node.data, end=" ")
        self.inorder(node.right)

    # Preorder Traversal (Root, Left, Right)
    def preorder(self, node):
        if not node:
            return
        print(node.data, end=" ")
        self.preorder(node.left)
        self.preorder(node.right)

    # Postorder Traversal (Left, Right, Root)
    def postorder(self, node):
        if not node:
            return
        self.postorder(node.left)
        self.postorder(node.right)
        print(node.data, end=" ")

    # Public wrappers
    def print_inorder(self):
        print("Inorder Traversal: ", end="")
        self.inorder(self.root)
        print()

    def print_preorder(self):
        print("Preorder Traversal: ", end="")
        self.preorder(self.root)
        print()

    def print_postorder(self):
        print("Postorder Traversal: ", end="")
        self.postorder(self.root)
        print()

# Driver code
if __name__ == "__main__":
    tree = Tree()
    s = input("Enter tree (level order, space separated): ")  # Example: 10 20 30 N 40 N N
    tree.build_tree(s)

    tree.print_inorder()
    tree.print_preorder()
    tree.print_postorder()
